### PR TITLE
apiserver/authconfig: make CEL compiler shareable

### PR DESCRIPTION
**Pull request by sttts**
/kind cleanup

#### What this PR does / why we need it:

The CEL compiler construction is quite heavy. It is done now during creation of every delegated authorizer although the compiler is static. We see it showing up in pprof in code that creates a number of authorizers. This PR makes the compiler passed down explicitly in authn/z config validation and added to the delegated authorizer configuration passed to the delegate authorizer constructor. This was higher level code can use a shared copy.

#### Which issue(s) this PR fixes:

Fixes #

#### Does this PR introduce a user-facing change?

```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

**Comment by sttts** [comment]
As changed, this is not compatible with `featuregatetesting.SetFeatureGateDuringTest` 🤔

**Comment by aojea** [comment]
/assign @jpbetz @cici37 

is there any side effect of reusing the compiler? performance, security , ...?

**Comment by jpbetz** [comment]
> /assign Joe Betz Cici Huang
> 
> is there any side effect of reusing the compiler? performance, security , ...?

This is safe, efficient and secure.  The compiler is really just a CEL environment wrapped in some conveniences.  The underlying CEL environment is designed to be shared by multiple compilation tasks. The environment defines what libraries, language settings and types are declared, and is thread-safe.

**Comment by jpbetz** [comment]
Deferring to @cici37 for review, but I didn't see anything concerning with a quick scan of the code.

**Review by cici37 — COMMENTED** [review]
The changes look good to me. The failing integration test is not expected though. LabelSelector should be supported with feature gate on and env version bumped.

File: pkg/kubeapiserver/authorizer/config.go [hunk 1]
```diff
@@ -172,7 +173,7 @@ func LoadAndValidateData(data []byte, requireNonWebhookTypes sets.Set[authzconfi
 	}
 
 	// validate the file and return any error
-	if errors := validation.ValidateAuthorizationConfiguration(nil, authorizationConfiguration,
+	if errors := validation.ValidateAuthorizationConfiguration(authorizationcel.DefaultCompiler(), nil, authorizationConfiguration,
```
Comment (liggitt) [inline pkg/kubeapiserver/authorizer/config.go:176]: we've had terrible trouble with cel envs getting locked by something called at init time before flags and feature gates are finalized... this was slightly untangled in https://github.com/kubernetes/kubernetes/pull/125571 (see the `[Move CEL env initialization out of package init()](https://github.com/kubernetes/kubernetes/pull/125571/commits/03d48b76831a3a02d503c3075d818a76afd83cd8)` commit) but I worry this is re-tangling things. The integration test related to this area failing in this PR doesn't give me confidence :-/

File: staging/src/k8s.io/apiserver/pkg/authentication/cel/compile.go [hunk 2]
```diff
@@ -39,6 +40,20 @@ type compiler struct {
 	varEnvs map[string]*environment.EnvSet
 }
 
+var (
+	defaultCompilerOnce sync.Once
+	defaultCompiler     Compiler
+)
+
+// DefaultCompiler returns a shared default Compiler. Do not modify the returned Compiler.
```
Comment (liggitt) [inline staging/src/k8s.io/apiserver/pkg/authentication/cel/compile.go:48]: this must not be called before feature gates and compatibility version are locked, right?

File: staging/src/k8s.io/apiserver/pkg/authorization/cel/compile.go [hunk 3]
```diff
@@ -65,6 +66,20 @@ type compiler struct {
 	envSet *environment.EnvSet
 }
 
+var (
+	defaultCompilerOnce sync.Once
+	defaultCompiler     Compiler
+)
+
+// DefaultCompiler returns a shared default Compiler. Do not modify the returned Compiler.
```
Comment (liggitt) [inline staging/src/k8s.io/apiserver/pkg/authorization/cel/compile.go:74]: same comment about not calling this before feature gates and compatibility versions are finalized? how are we guarding against that?

File: staging/src/k8s.io/apiserver/pkg/authentication/cel/compile.go (hunk 2 above)
Comment (sttts) [reply staging/src/k8s.io/apiserver/pkg/authentication/cel/compile.go:48]: right

File: staging/src/k8s.io/apiserver/pkg/authorization/cel/compile.go (hunk 3 above)
Comment (sttts) [reply staging/src/k8s.io/apiserver/pkg/authorization/cel/compile.go:74]: see https://github.com/kubernetes/kubernetes/pull/127504#discussion_r1774737883, then it is by construction.

File: staging/src/k8s.io/apiserver/pkg/apis/apiserver/validation/validation.go [hunk 4]
```diff
@@ -616,7 +613,7 @@ func compileUserCELExpression(compiler authenticationcel.Compiler, expression au
 }
 
 // ValidateAuthorizationConfiguration validates a given AuthorizationConfiguration.
-func ValidateAuthorizationConfiguration(fldPath *field.Path, c *api.AuthorizationConfiguration, knownTypes sets.String, repeatableTypes sets.String) field.ErrorList {
+func ValidateAuthorizationConfiguration(compiler authorizationcel.Compiler, fldPath *field.Path, c *api.AuthorizationConfiguration, knownTypes sets.String, repeatableTypes sets.String) field.ErrorList {
```
[the remark this answers is not in the document]
Comment (sttts) [reply staging/src/k8s.io/apiserver/pkg/apis/apiserver/validation/validation.go:616]: fixed in extra commit.

**Comment by sttts** [comment]
@liggitt have rewired: compiler is now explicit in validation code and delegated authorizer construction. It is recreated in the higher level kube-apiserver plumbing and apiserver options-related code. Seems to be the best of both worlds. Those who instantiate multiple delegated authorizers can do so with a shared instance.

File: pkg/kubeapiserver/options/authentication.go [hunk 5]
```diff
@@ -574,7 +575,7 @@ func (o *BuiltInAuthenticationOptions) ToAuthenticationConfig() (kubeauthenticat
 		}
 	}
 
-	if err := apiservervalidation.ValidateAuthenticationConfiguration(ret.AuthenticationConfig, ret.ServiceAccountIssuers).ToAggregate(); err != nil {
+	if err := apiservervalidation.ValidateAuthenticationConfiguration(authenticationcel.NewDefaultCompiler(), ret.AuthenticationConfig, ret.ServiceAccountIssuers).ToAggregate(); err != nil {
```
Comment (liggitt) [inline pkg/kubeapiserver/options/authentication.go:578]: it's not clear to me what principle is used to decide whether to plumb in a configurable compiler (like staging/src/k8s.io/apiserver/plugin/pkg/authenticator/token/oidc/oidc.go does) and what chooses to call the global default constructor, like this does ... how did you decide which to do in each call site?

File: pkg/kubeapiserver/options/authentication.go [hunk 6]
```diff
@@ -756,7 +757,7 @@ func (o *BuiltInAuthenticationOptions) ApplyTo(
 					return
 				}
 
-				validationErrs := apiservervalidation.ValidateAuthenticationConfiguration(authConfig, authenticatorConfig.ServiceAccountIssuers)
+				validationErrs := apiservervalidation.ValidateAuthenticationConfiguration(authenticationcel.NewDefaultCompiler(), authConfig, authenticatorConfig.ServiceAccountIssuers)
```
Comment (liggitt) [inline pkg/kubeapiserver/options/authentication.go:760]: same question about constructing new or plumbing down ... not sure how to decide which to do here

File: pkg/kubeapiserver/authorizer/config.go [hunk 7]
```diff
@@ -172,9 +174,9 @@ func LoadAndValidateData(data []byte, requireNonWebhookTypes sets.Set[authzconfi
 	}
 
 	// validate the file and return any error
-	if errors := validation.ValidateAuthorizationConfiguration(nil, authorizationConfiguration,
-		sets.NewString(modes.AuthorizationModeChoices...),
-		sets.NewString(repeatableAuthorizerTypes...),
+	if errors := validation.ValidateAuthorizationConfiguration(authorizationcel.NewDefaultCompiler(), nil, authorizationConfiguration,
```
Comment (liggitt) [inline pkg/kubeapiserver/authorizer/config.go:177]: same question about whether to plumb or construct new ... it seems confusing to use one compiler to validate the config and a different compiler (held by reloadableAuthorizerResolver) to construct the authorizer

File: pkg/kubeapiserver/authorizer/reload.go [hunk 8]
```diff
@@ -61,6 +61,7 @@ type reloadableAuthorizerResolver struct {
 	nodeAuthorizer *node.NodeAuthorizer
 	rbacAuthorizer *rbac.RBACAuthorizer
 	abacAuthorizer abac.PolicyList
+	compiler       authorizationcel.Compiler
```
Comment (liggitt) [inline pkg/kubeapiserver/authorizer/reload.go:64]: document where this compiler should be obtained from? is it required to be non-nil?

File: staging/src/k8s.io/apiserver/plugin/pkg/authenticator/token/oidc/oidc.go [hunk 9]
```diff
@@ -86,6 +86,9 @@ type Options struct {
 	// Optional http.Client used to make all requests to the remote issuer.  Mutually exclusive with CAContentProvider.
 	Client *http.Client
 
+	// Opitional CEL compiler used to compile the CEL expressions. If not provided, a default compiler will be used.
```
[the remark this answers is not in the document]
Comment (sttts) [reply staging/src/k8s.io/apiserver/plugin/pkg/authenticator/token/oidc/oidc.go:89]: fixed

File: staging/src/k8s.io/apiserver/plugin/pkg/authenticator/token/oidc/oidc.go [hunk 10]
```diff
@@ -86,6 +86,9 @@ type Options struct {
...
+	Compiler authenticationcel.Compiler
```
[the remark this answers is not in the document]
Comment (sttts) [reply staging/src/k8s.io/apiserver/plugin/pkg/authenticator/token/oidc/oidc.go:90]: This is not a generic compiler, but the interface from the authentication cel package. There are these two constructors:
```
// NewDefaultCompiler returns a new Compiler following the default compatibility version.
func NewDefaultCompiler() Compiler {
	return NewCompiler(environment.MustBaseEnvSet(environment.DefaultCompatibilityVersion(), true))
}

// NewCompiler returns a new Compiler.
func NewCompiler(env *environment.EnvSet) Compiler {
	return &compiler{
		varEnvs: mustBuildEnvs(env),
	}
}
```
So it should be clear what to pass there. I have added more docs why this field exists and when to use it.

File: staging/src/k8s.io/apiserver/pkg/authorization/cel/compile.go (hunk 3 above)
Comment (sttts) [reply staging/src/k8s.io/apiserver/pkg/authorization/cel/compile.go]: Adding a note everywhere where it depends on feature gates.

File: staging/src/k8s.io/apiserver/pkg/authorization/cel/compile.go [hunk 11]
```diff
@@ -65,6 +65,11 @@ type compiler struct {
 	envSet *environment.EnvSet
 }
 
+// NewDefaultCompiler returns a new Compiler following the default compatibility version.
+func NewDefaultCompiler() Compiler {
+	return NewCompiler(environment.MustBaseEnvSet(environment.DefaultCompatibilityVersion(), true))
```
[the remark this answers is not in the document]
Comment (sttts) [reply staging/src/k8s.io/apiserver/pkg/authorization/cel/compile.go:70]: Done:
```
// Note: the compiler construction depends on feature gates and the compatibility version to be initialized.
```

File: staging/src/k8s.io/apiserver/pkg/authentication/cel/compile.go [hunk 12]
```diff
@@ -39,6 +39,11 @@ type compiler struct {
 	varEnvs map[string]*environment.EnvSet
 }
 
+// NewDefaultCompiler returns a new Compiler following the default compatibility version.
```
[the remark this answers is not in the document]
Comment (sttts) [reply staging/src/k8s.io/apiserver/pkg/authentication/cel/compile.go:42]: I wonder, we probably should come up with a mechanism that detects that, at least in CI. Like changing the gates and compatiblity version after a consumption should fail, with some way to disable that mechanism in tests.

File: pkg/kubeapiserver/options/authentication.go (hunk 5 above)
Comment (sttts) [reply pkg/kubeapiserver/options/authentication.go:578]: Validation is stateless and hence gets the argument, always. Here I am considering the compiler with its heavy initialization as state, i.e. it belongs onto the heap, not the stack.

Constructors like for oidc or webhooks return state, so they are in general fine with construction of another compiler. But I have made it shareable/configurable for those constructors that showed up in pprof.

File: pkg/kubeapiserver/options/authentication.go (hunk 6 above)
Comment (sttts) [reply pkg/kubeapiserver/options/authentication.go:760]: Applying options is a one-time operation. Hence, not creating a new one is fine.

File: pkg/kubeapiserver/authorizer/reload.go (hunk 8 above)
Comment (sttts) [reply pkg/kubeapiserver/authorizer/reload.go:64]: done

Comment (sttts) [reply pkg/kubeapiserver/authorizer/reload.go:64]: Note: here we could make it configurable from outside through `Config`. But as this is non-library code of kube-apiserver (by the package name), I have not done that.

**Comment by sttts** [comment]
Addressed comments.

File: pkg/volume/plugins_test.go [hunk 13]
```diff
@@ -198,14 +198,14 @@ func TestVolumePluginMultiThreaded(t *testing.T) {
 	assert.Equal(t, int32(0), totalErrors.Load())
 
 	for i := 0; i < 100; i++ {
+		wg.Add(1)
```
Comment (liggitt) [inline pkg/volume/plugins_test.go:201]: 🤦

File: staging/src/k8s.io/apiserver/pkg/authentication/cel/compile.go (hunk 12 above)
[the remark this answers is not in the document]
Comment (liggitt) [reply staging/src/k8s.io/apiserver/pkg/authentication/cel/compile.go:42]: sounds good... @jpbetz, mind opening an issue to track this?

**Pull request merged**